### ISSUE TITLE
Diagonal does not return same size due to broadcast

### ISSUE THREAD

**Issue by janEbert** [FluxML/Flux.jl#890]
If you call a `Diagonal` on an `Array` whose first dimension is `1`, that dimension will be expanded to the size of `Diagonal` due to broadcasting.
Example:
```julia
julia> using Flux

julia> diag = Flux.Diagonal(5)
Diagonal(5)

julia> size(diag(rand(1, 5, 2))  # expected is (1, 5, 2)
(5, 5, 2)
```

This can be avoided if we `dropdims` all the first dimensions that are equal to one.
The dropped dimensions could be counted and later `Flux.unsqueeze`d to the front again.

**Comment by janEbert** [comment]
I fixed it using the following code:
```julia
function apply_layer_norm(layer_norm::LayerNorm, x)
    size(x, 1) != 1 && return layer_norm(x)
    return reshape(selectdim(layer_norm(x), 1, 1), size(x))
end
```

If desired, I can submit a PR which modifies the `(diag::Diagonal)(x)` function accordingly, making it a bit more ugly for handling this special case.

**Comment by bhvieira** [comment]
I think it should error out if `size(x,1)` and `in` are different. At least, that's what the docs imply.

**Comment by janEbert** [comment]
I don't think this was documented when I opened the issue. I'll close this then, thanks. :)

**Comment by bhvieira** [comment]
We still need to implement the error though. I know Diagonal is used in LayerNorm, perhaps there's some implication as well.

**Comment by CarloLucibello** [comment]
we don't have a Diagonal Layer anymore
